server/src/main/java/org/opensearch/index/mapper/FlatObjectFieldMapper.java

```diff
@@ -0,0 +1,745 @@
+            return new PrefixQuery(new Term(directSubfield, indexedValueForSearch(rewriteValue)), method);
+        }
+
+        @Override
+        public Query rangeQuery(Object lowerTerm, Object upperTerm, boolean includeLower, boolean includeUpper, QueryShardContext context) {
+            String directSubfield = directSubfield();
+            String rewriteUpperTerm = rewriteValue(((BytesRef) upperTerm).utf8ToString());
+            String rewriteLowerTerm = rewriteValue(((BytesRef) lowerTerm).utf8ToString());
+            if (context.allowExpensiveQueries() == false) {
+                throw new OpenSearchException(
+                    "[range] queries on [text] or [keyword] fields cannot be executed when '"
+                        + ALLOW_EXPENSIVE_QUERIES.getKey()
+                        + "' is set to false."
+                );
+            }
+            failIfNotIndexed();
+            return new TermRangeQuery(
+                directSubfield,
+                lowerTerm == null ? null : indexedValueForSearch(rewriteLowerTerm),
+                upperTerm == null ? null : indexedValueForSearch(rewriteUpperTerm),
+                includeLower,
+                includeUpper
+            );
+        }
+
+        /**
+         * if there is dot path. query the field name in flatObject parent field.
+         * else query in _field_names system field
+         */
+        @Override
+        public Query existsQuery(QueryShardContext context) {
+            String searchKey;
+            String searchField;
+            if (name().contains(DOT_SYMBOL)) {
+                searchKey = name().split("\\.")[0];
+                searchField = name();
+            } else {
+                searchKey = FieldNamesFieldMapper.NAME;
+                searchField = name();
+            }
+            return new TermQuery(new Term(searchKey, indexedValueForSearch(searchField)));
+        }
+
+        @Override
+        public Query wildcardQuery(
+            String value,
+            @Nullable MultiTermQuery.RewriteMethod method,
+            boolean caseInsensitve,
+            QueryShardContext context
+        ) {
+            // flat_object field types are always normalized, so ignore case sensitivity and force normalize the wildcard
+            // query text
+            throw new QueryShardException(
+                context,
+                "Can only use wildcard queries on keyword and text fields - not on [" + name() + "] which is of type [" + typeName() + "]"
+            );
+        }
+
+    }
+
+    private final ValueFieldMapper valueFieldMapper;
+    private final ValueAndPathFieldMapper valueAndPathFieldMapper;
+    private final IndexAnalyzers indexAnalyzers;
+
+    FlatObjectFieldMapper(
+        String simpleName,
+        FieldType fieldType,
+        FlatObjectFieldType mappedFieldType,
+        ValueFieldMapper valueFieldMapper,
+        ValueAndPathFieldMapper valueAndPathFieldMapper,
+        MultiFields multiFields,
```

Comment:
multiFields is never used -> remove it

Reply:
removed in the commit @mingshl
[Removed license head and add tests for wildcard query](https://github.com/opensearch-project/OpenSearch/pull/6507/commits/83a6b09af349fdddb48c206ecd049e1f72c92d10)